### ISSUE TITLE
Wrong parts selected and grouped

### ISSUE THREAD

**Issue by jrzwee** [openscopeproject/InteractiveHtmlBom#408]
When exporting the iBom with the command line on the latest EAGLE version, I get this output:

![image](https://github.com/openscopeproject/InteractiveHtmlBom/assets/10603792/6ee1982e-9d90-4e64-8110-f99d169f1be0)

Random parts are grouped and if I click on one part in the viewer it is the wrong part selected.

E.g.: I click on an IC and the table tells me it is a resistor.

**Comment by qu1ck** [comment]
Can you provide an example board that shows the issue?

**Comment by Funkenjaeger** [comment]
I'm happy to take a look as soon as @jrzwee can provide an example .brd file that reproduces the issue.

**Comment by qu1ck** [comment]
Thanks, I see the issue with this board. Haven't delved deep yet, but this is not a recently introduced bug at least, probably eagle parser doesn't handle some case.

**Comment by Funkenjaeger** [comment]
Thanks for sharing the example, I'll take a look.  I wouldn't be surprised if it's related to assembly variants, that's a feature I haven't used much so there may well be some surprises that the parser doesn't yet handle.

**Comment by Funkenjaeger** [comment]
Okay, so the problem was related to assembly variants, but only tangentially.  Assembly variants are the only way to natively accomplish DNP in Eagle/Fusion 360, and the way I was handling those DNP's was wrong, resulting in the indexing of components to footprints being all messed up, thus the incorrect highlighting.  PR incoming.

**Comment by Funkenjaeger** [comment]
Submitted #409 

@jrzwee can you please give it a try from my fork and confirm whether it's fixed this issue on your real board?

**Comment by jrzwee** [comment]
You guys are great! I can confirm it is working now! @Funkenjaeger

**Comment by jrzwee** [comment]
What is still a problem: the copper is missing when exported.

**Comment by qu1ck** [comment]
It is off by default. Add `--include-tracks` and you probably also want `--include-nets` flags to your command.

**Comment by qu1ck** [comment]
@Funkenjaeger on the test board above generating zones appears a bit broken, they don't have any cutouts for traces/vias etc. I can't recall if they were ever fully supported, is there a bug or something preventing correct holes processing?

**Comment by jrzwee** [comment]
> It is off by default. Add `--include-tracks` and you probably also want `--include-nets` flags to your command.

Unfortunately adding the flags seems to lead to the same result.

**Comment by Funkenjaeger** [comment]
> @Funkenjaeger on the test board above generating zones appears a bit broken, they don't have any cutouts for traces/vias etc. I can't recall if they were ever fully supported, is there a bug or something preventing correct holes processing?

Unfortunately that's due to the fact that Eagle handles zone cutouts internally while rendering, so the geometry of the cutouts is not embedded in the BRD file, and therefore it goes beyond merely parsing the file.  What's in the file is just the geometry of the zone (sans cutouts) and the set of design rule parameters which it uses to determine the cutouts.  
In fact, evidencing this, when you first open up a BRD in Eagle, the zones are rendered as unfilled polygons with a dashed outline:
<img width="284" alt="image" src="https://github.com/openscopeproject/InteractiveHtmlBom/assets/24237058/3cd39b40-aa3d-412b-a451-330f9238191b">
And only after hitting 'Ratsnest' does it actually calculate and render the copper:
<img width="405" alt="image" src="https://github.com/openscopeproject/InteractiveHtmlBom/assets/24237058/4d38c49c-d133-458d-b38f-aa8276d0c0ce">

So, I think to address this I'd either have to do it in the parser (implement similar logic as Eagle does in its rendering, transform each zone into one or more much more complex polygons), or do it during rendering (in javascript).  The former seems like it would be a huge headache, the latter seems maybe feasible (composite operations in canvas?) but it would mean Eagle-specific changes residing outside the Eagle parser.  Of course, I welcome your thoughts on whether there's an easier way, etc.  We could put it in as a wishlist/enhancement issue if you want to track it - it's not something I would necessarily rush to tackle unless other users are asking for it, due to the likely amount of effort involved.

**Comment by Funkenjaeger** [comment]
> > It is off by default. Add `--include-tracks` and you probably also want `--include-nets` flags to your command.
> 
> Unfortunately adding the flags seems to lead to the same result.

Just to be clear, are you saying you're not seeing copper at all, or that the copper isn't correct (e.g. due to the limitation with cutouts in zones just mentioned)?

This is what I get with `--include-tracks` and `--include-nets` - note that all the other copper residing inside the large zones (copper pours) are present and can be highlighted by net as usual, despite the cutouts limitation:
<img width="847" alt="image" src="https://github.com/openscopeproject/InteractiveHtmlBom/assets/24237058/ffb3aadf-204d-4675-86b8-d00407545193">

If you're not seeing copper at all, are you sure that in addition to the flags, you also enabled display of Tracks and Zones in the options menu in the generated HTML file?

**Comment by qu1ck** [comment]
> So, I think to address this I'd either have to do it in the parser (implement similar logic as Eagle does in its rendering, transform each zone into one or more much more complex polygons), or do it during rendering (in javascript). The former seems like it would be a huge headache, the latter seems maybe feasible (composite operations in canvas?) but it would mean Eagle-specific changes residing outside the Eagle parser. Of course, I welcome your thoughts on whether there's an easier way, etc. We could put it in as a wishlist/enhancement issue if you want to track it - it's not something I would necessarily rush to tackle unless other users are asking for it, due to the likely amount of effort involved.

Doing this in javascript is not feasible, no canvas operations will possibly match EDA specific things like hatched zones, spike connections, corner rounding, netclass clearance rules and all other details that every package implements slightly differently.

I think it should be implemented in python or zones should be skipped all together, like in EasyEDA parser. Maybe proper zone support can be a differentiating feature of the eagle->generic json->ibom workflow if you are still supporting that.

> Unfortunately adding the flags seems to lead to the same result.

Check html settings, either reset them or enable tracks manually.

**Comment by Funkenjaeger** [comment]
> Doing this in javascript is not feasible, no canvas operations will possibly match EDA specific things like hatched zones, spike connections, corner rounding, netclass clearance rules and all other details that every package implements slightly differently.

And for the same reasons I won't be able to get it perfectly right in the parser either.  My cursory thought was that if I were to attempt it at all, I'd only punch cutouts around traces/pads/vias naively using global clearance rules and not handle all the other, more nuanced details like those you mention - thinking that it would be 'close enough' for the general ibom use case of assembly/troubleshooting aid.

> I think it should be implemented in python or zones should be skipped all together, like in EasyEDA parser. Maybe proper zone support can be a differentiating feature of the eagle->generic json->ibom workflow if you are still supporting that.

I can drop zone processing from the parser.  
I can still support the genericjson workflow as long as Autodesk doesn't break it, like they did (and subsequently un-did) a while back.

**Comment by qu1ck** [comment]
I would say "good enough" to be better than no zone support at all is having following:
* Having correct clearance to other nets (taking into account both global clearance and net to net clearance if specified)
* Having connections to items of same net, simple solid connection, ignoring spikes is fine
* Having any manual cutouts (i.e. keepout areas)
Not necessary:
* Spiked connections
* Hatching
* Corner rounding
* Min width and any other rarely used rules

I understand that even with these simplifications this is not a simple task and likely needs a 3rd party geometry library for all the polygon cutting/padding/fracturing operations.

I mentioned the ULP workflow in case it gives you access to the calculated zone polygon that is not stored in the board thus simplifying all the work. But I don't know if that is the case.

**Comment by Funkenjaeger** [comment]
Copy that, makes sense.  I think my path forward is to remove zone processing for now and put in a wishlist/enhancement issue to capture the possible future addition of all that geometry processing in python.

> I mentioned the ULP workflow in case it gives you access to the calculated zone polygon that is not stored in the board thus simplifying all the work. But I don't know if that is the case.

It does, fortunately, and that workflow currently handles cutouts as a result.

**Comment by jrzwee** [comment]
Here you can see the output I got in 2021 with the brd2json EAGLE ULP and the output I got with https://github.com/openscopeproject/InteractiveHtmlBom/pull/409 .

**2021 brd2json**
![msedge_orVs4ovfel](https://github.com/openscopeproject/InteractiveHtmlBom/assets/10603792/b8d4c81b-5cfe-4cd2-b37c-d5d67406556f)

**2023**
![msedge_zstAaJRjOO](https://github.com/openscopeproject/InteractiveHtmlBom/assets/10603792/699e01ab-1e9b-4c04-819b-7437f80847b8)

**Comment by qu1ck** [comment]
@jrzwee Yeah, that's what we were discussing above. In your second picture tracks are there, they are just covered by the zone because it doesn't have cutouts. If you care about correct rendering of zones use brd2json.

**Comment by qu1ck** [comment]
Closing since the main issue was fixed. @Funkenjaeger you can open another issue for tracking disabling zones if you want or just send an MR when ready.
